fixes package for Meteor 2.3 and Meteor 3.0

drone1 opened the pull request:
passes all tests for meteor 1.12.1, 2.3, 3.0.3

drone1 committed:
fixes package for Meteor 2.3 and Meteor 3.0

drone1 committed:
adds tests for meteor 3+

drone1 committed:
bump version from 1.1.0 => 1.2.0

dburles commented on package.js:
```diff
@@ -1,13 +1,13 @@
 
 Package.onUse(function(api) {
-  api.versionsFrom('1.4.2');
+  api.versionsFrom(['1.12.1', '2.3', '3.0']);
```
Hey @drone1 just wondering why the minimum version bump here is required?

drone1 replied:
Sorry, actually it looks like you're referring to the package minimum version but you replied on this line about minimum Meteor versions. If so, I bumped the package version since there is an update? Sorry I'm not understanding the question. Thank you.

dburles replied:
What is the specific trouble you were having running the current version of the package on Meteor 3? It works okay for me.

dburles replied:
What I meant here is that you have dropped support for Meteor versions between 1.4.2 and 1.12.1 and I was wondering if there was any specific reason for that?

drone1 replied:
> What is the specific trouble you were having running the current version of the package on Meteor 3? It works okay for me.

Helpers weren't working on a client collection on document retrieved with `findOne()`. Once I changed the `api.versionFrom` line everything worked again from what I recall, but let me confirm and get back here with details.

> What I meant here is that you have dropped support for Meteor versions between 1.4.2 and 1.12.1 and I was wondering if there was any specific reason for that?

This was totally copy-pasta :( Will fix. Thanks.

dburles replied:
I can't reproduce the problem you mention, maybe something else is going on?

drone1 commented:
Hey! Thanks for the reply.

To be honest I copied this from aldeed:collections2, which works with
modern versions of Meteor, and it fixed it – I’m not sure what it means to
have three minimum versions.

On Sun 29. Sep 2024 at 22:57, David Burles ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In package.js
> <https://github.com/dburles/meteor-collection-helpers/pull/88#discussion_r1780170032>
> :
>
> >    documentation: "README.md",
>    git: "https://github.com/dburles/meteor-collection-helpers.git",
>  });
>
>  Package.onUse(function(api) {
> -  api.versionsFrom('1.4.2');
> +  api.versionsFrom(['1.12.1', '2.3', '3.0']);
>
> Hey @drone1 <https://github.com/drone1> just wondering why the minimum
> version bump here is required?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#pullrequestreview-2335997014>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPAWY6A7QI5CC52CZCLZZBSUDAVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGMZVHE4TOMBRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drone1 committed:
fixes my copy-paste error in which i replaced the old `api.versionsFrom()` version (in dburles:collection-helpers@1.1.0) of 1.4.2 with 1.12.1. this change corrects this (back to 1.4.2) but also retains the fix for meteor 3, which passes 3 versions rather than just a single version, and includes 2.3 and 3.0.

Tests are passing for Meteor 2.3 and 3.0, using commands:

```
  meteor test-packages ./packages/collection-helpers --release 2.3
  meteor test-packages ./packages/collection-helpers --release 3.0
```

Meteor 1.4.2 is spewing some errors about a failure to "get
local issuer certificate" (macOS Sonoma 14.5), but I wanted to push this
commit in case author @dburles wants to try and test.

drone1 commented:
In a Meteor 3.0.3 project, if I try to add my local package (with the corrected `  api.versionsFrom('1.4.2')` line), combined with some other common packages (see below) installed, I get this error:
```
   While selecting package versions:
   error: Conflict: Constraint mongo@1.1.14 is not satisfied by mongo 2.0.2.
   Constraints on package "mongo":
   * mongo@2.0.2 <- top level
   * mongo@~2.0.2 <- top level
   * mongo@2.0.0 <- reactive-dict 1.3.2 <- ostrio:flow-router-extra
   3.11.0-rc300.1
   * mongo@1.1.14 <- dburles:collection-helpers 1.0.0
   * mongo@1.10.1 || 1.12.0 || 2.0.0 <- aldeed:collection2 4.0.3
   * mongo@2.0.2 <- accounts-base 3.0.2
   * mongo@1.16.1 || 2.0.0 <- alanning:roles 4.0.0   
```

However, if I change the line to `api.versionsFrom(['1.4.2', '2.3', '3.0'])` (NOTE: I fixed the issue you mentioned in [this comment](https://github.com/dburles/meteor-collection-helpers/pull/88#discussion_r1781909310)), no errors.

Actually, I can't even get into a state where I can test the original problem I was having, because I cannot install versions 1.0.0 or 1.1.0 to my project, even with `--allow-incompatible-update`. (Reminder: the original problem was that helpers simply did not exist on the client.)

But yes, **tests pass for Meteor 2.3 and 3.0.**

However I can't get up and running with Meteor 1.4.2:

```
> meteor test-packages ./packages/collection-helpers --release 1.4.2

[[[[[ Tests ]]]]]

=> Started proxy.
=> Started MongoDB.
=> Errors prevented startup:

   While downloading allow-deny@1.0.9...:
   error: unable to get local issuer certificate

   While downloading autoupdate@1.3.12...:
   error: unable to get local issuer certificate

   While downloading babel-compiler@6.24.7...:
   error: unable to get local issuer certificate

   While downloading babel-runtime@1.1.1...:
   error: unable to get local issuer certificate

   While downloading base64@1.0.10...:
   error: unable to get local issuer certificate

   While downloading binary-heap@1.0.10...:
   error: unable to get local issuer certificate

   While downloading blaze@2.1.9...:
   error: unable to get local issuer certificate

   While downloading boilerplate-generator@1.3.1...:
   error: unable to get local issuer certificate

   While downloading bootstrap@1.0.1...:
   error: unable to get local issuer certificate

   While downloading caching-compiler@1.0.6...:
   error: unable to get local issuer certificate

   While downloading caching-html-compiler@1.0.6...:
   error: unable to get local issuer certificate

   While downloading callback-hook@1.0.10...:
   error: unable to get local issuer certificate

   While downloading check@1.2.5...:
   error: unable to get local issuer certificate

   While downloading ddp-client@2.2.0...:
   error: unable to get local issuer certificate

   While downloading ddp-common@1.3.0...:
   error: unable to get local issuer certificate

   While downloading ddp-server@2.1.1...:
   error: unable to get local issuer certificate

   While downloading diff-sequence@1.0.7...:
   error: unable to get local issuer certificate

   While downloading ecmascript@0.9.0...:
   error: unable to get local issuer certificate

   While downloading ecmascript-runtime@0.5.0...:
   error: unable to get local issuer certificate

   While downloading ecmascript-runtime-client@0.5.0...:
   error: unable to get local issuer certificate

   While downloading ecmascript-runtime-server@0.5.0...:
   error: unable to get local issuer certificate

   While downloading ejson@1.1.0...:
   error: unable to get local issuer certificate

   While downloading http@1.2.10...:
   error: unable to get local issuer certificate

   While downloading id-map@1.0.9...:
   error: unable to get local issuer certificate

   While downloading jquery@1.11.11...:
   error: unable to get local issuer certificate

   While downloading logging@1.1.19...:
   error: unable to get local issuer certificate

   While downloading meteor@1.8.6...:
   error: unable to get local issuer certificate

   While downloading minifier-js@1.1.13...:
   error: unable to get local issuer certificate

   While downloading minimongo@1.4.4...:
   error: unable to get local issuer certificate

   While downloading modules@0.11.8...:
   error: unable to get local issuer certificate

   While downloading modules-runtime@0.9.2...:
   error: unable to get local issuer certificate

   While downloading mongo@1.1.24...:
   error: unable to get local issuer certificate

   While downloading mongo-id@1.0.6...:
   error: unable to get local issuer certificate

   While downloading npm-mongo@2.2.34...:
   error: unable to get local issuer certificate

   While downloading ordered-dict@1.0.9...:
   error: unable to get local issuer certificate

   While downloading promise@0.10.2...:
   error: unable to get local issuer certificate

   While downloading random@1.0.10...:
   error: unable to get local issuer certificate

   While downloading reactive-dict@1.1.9...:
   error: unable to get local issuer certificate

   While downloading reload@1.1.11...:
   error: unable to get local issuer certificate

   While downloading retry@1.0.9...:
   error: unable to get local issuer certificate

   While downloading routepolicy@1.0.12...:
   error: unable to get local issuer certificate

   While downloading session@1.1.7...:
   error: unable to get local issuer certificate

   While downloading spacebars@1.0.13...:
   error: unable to get local issuer certificate

   While downloading spacebars-compiler@1.0.12...:
   error: unable to get local issuer certificate

   While downloading templating@1.2.14_1...:
   error: unable to get local issuer certificate

   While downloading templating-tools@1.0.4...:
   error: unable to get local issuer certificate

   While downloading test-in-browser@1.1.0...:
   error: unable to get local issuer certificate

   While downloading tinytest@1.0.12...:
   error: unable to get local issuer certificate

   While downloading tracker@1.1.3...:
   error: unable to get local issuer certificate

   While downloading url@1.0.11...:
   error: unable to get local issuer certificate

   While downloading webapp@1.4.0...:
   error: unable to get local issuer certificate

=> Your application has errors. Waiting for file change.
```

`.meteor/packages`:
```
meteor-base@1.5.2             # Packages every Meteor app needs to have
mongo@2.0.2                   # The database Meteor supports right now
jquery                  # Wrapper package for npm-installed jquery

standard-minifier-css@1.9.3   # CSS minifier run for production mode
standard-minifier-js@3.0.0    # JS minifier run for production mode
es5-shim@4.8.1                # ECMAScript 5 compatibility for older browsers
ecmascript@0.16.9              # Enable ECMAScript2015+ syntax in app code

hot-module-replacement@0.5.4  # Update code in development without reloading the page

dburles:collection-helpers@1.2.0
```

I tried running with `METEOR_NO_HTTPS=1` but no luck. Have you seen these errors before? Any suggestions?

Thanks.

dburles commented:
Hey thanks for taking the time to test on earlier versions of Meteor. I'm happy with the version changes. Did you want to investigate async helpers?

drone1 commented:
Ok great!

Yeah, I can try and take a look at some point and submit another PR. I’ve
needed async helpers myself now with the recent meteor updates…

On Tue 15. Oct 2024 at 23:58, David Burles ***@***.***> wrote:

> Hey thanks for taking the time to test on earlier versions of Meteor. I'm
> happy with the version changes. Did you want to investigate async helpers?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#issuecomment-2415237506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPHVFTK3QYMW7ZEJFZTZ3WFYVAVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJVGIZTONJQGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dburles commented:
Great. I don't imagine it would require any changes to the source to support them. It's more a matter of including some tests for Meteor 3.

drone1 commented:
Good point. Let me see what I can do.

On Wed, Oct 16, 2024 at 1:06 AM David Burles ***@***.***>
wrote:

> Great. I don't imagine it would require any changes to the source to
> support them. It's more a matter of including some tests for Meteor 3.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#issuecomment-2415317336>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPG4CIRKSDIAF5VCMKDZ3WNYBAVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJVGMYTOMZTGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mcorbelli commented:
Any news?

drone1 commented:
Haven’t played with this yet. Are you waiting for these changes to accept
the PR? For some reason I was under the impression we were talking about a
secondary PR.

On Mon 28. Oct 2024 at 23:30, Mattia Corbelli ***@***.***>
wrote:

> Any news?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#issuecomment-2442794857>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPHZPGWVUAEVQK5ZLA3Z52UHJAVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBSG44TIOBVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drone1 commented:
Ah sure, let me take a look…

On Wed 30. Oct 2024 at 08:49, David Burles ***@***.***> wrote:

> I don't have any idea on how Meteor 3 works with non async queries. Where
> you have commented out a test with a todo it would be good to add some
> async helpers.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#issuecomment-2446089524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPBXTXZR5A72CLG3U63Z6CFPRAVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBWGA4DSNJSGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drone1 committed:
adds test for async helper, for meteor 3+

drone1 committed:
run tests on both the client and server

drone1 commented:
Can't seem to get tests running on the server, despite a variety of attempts. Do you have any experience with this? I ended up [posting on the Meteor forums](https://forums.meteor.com/t/meteor-test-packages-doesnt-run-tinytests-on-server/62498) but haven't heard anything yet. If you've got a tip, please let me know!

drone1 commented:
Should be all good here.

dburles commented on collection-helpers_tests.js:
```diff
@@ -1,60 +1,130 @@
+    test.equal(author.fullName(), 'Carl Sagan');
+
+    author = await Authors.findOneAsync(author1);
+    books = author.books();
+    test.equal(books.count(), 1);
```
Shouldn't this be failing since the helper is using the non async `Books.find`?

drone1 replied:
On the client no, but on the server, yes. Good catch. I'm guessing the returned `Promise` autocasts to a truthy value and so the comparison to ``1`` passes? I'll take a look.

drone1 replied:
Holllyyy cow, so if I put:
```
test.isTrue('x' === 'y')
```
in the Meteor 3+ tests and all tests pass, so long as I put that line AFTER the first 'await'.

I did a bit of digging and Tinytest does NOT play nice with `async`.

I'm going to look at replacing the tests with Mocha, I guess? Agg, kill me.

drone1 committed:
tinytest => puppeteer/mocha; tinytest doesn't like async code; copied aldeed:collection2 package structure and fixed up tests, which had at least one issue (thank you @dburles)

drone1 commented:
Updated with some structural changes and pushed, but need more testing and changes, possibly. Bear with me please. Thanks.

drone1 committed:
updates README with async examples & testing instructions

drone1 committed:
adds missing '#' for Meteor.users section

drone1 committed:
fixes testing/license heading level

drone1 commented:
@dburles I think we're good here. Please take a look when you can. Thanks.

dburles commented:
Hey @drone1 thanks for all your effort on this. I would like to keep Tinytest in place since it retains the existing file structure and simplicity. I'm going to open a new PR and cherry pick some of your commits over there.

dburles commented:
Hey @drone1 see PR #88

drone1 commented:
> Hey @drone1 thanks for all your effort on this. I would like to keep Tinytest in place since it retains the existing file structure and simplicity. I'm going to open a new PR and cherry pick some of your commits over there.

tinytest does not play nice with async as I mentioned [here](https://github.com/dburles/meteor-collection-helpers/pull/88/commits/3b066f2a719d4084c939c82a60c1e9297e1388ff) – the test may run, but it's likely not actually testing much. You can look into it, but it's completely outdated and supports only one async call per test, IIRC, which is why I moved over to a more modern environment.

I understand the resistance to the structural changes, but it's what [collection2](https://github.com/Meteor-Community-Packages/meteor-collection2) uses, and it works fine. Do you have a specific concern?

I put a lot more energy into this than I would have liked. It would be nice to not have it thrown away. It would also be nice to get the credit for the work. If there are specific concerns please let me know. Thanks a lot.

dburles commented:
Hey Tinytest is okay with async tests, it just requires that you use `Tinytest.addAsync` and call the `onComplete` function when you're done. See: https://github.com/meteor/meteor/tree/devel/packages/tinytest#synchronous-and-asynchronous-tests. If Tinytest becomes deprecated in future Meteor releases it will make sense to migrate. The test requirements for this package are pretty simple. I understand that you would appreciate credit and the your work isn't being thrown away as I have kept the changes to the tests you made, as well as the readme updates. See: https://github.com/dburles/meteor-collection-helpers/pull/89/commits/9085e8b695c2e73e896f7acbdd08aaac46882073. I really do appreciate the time you have put into working on this!

dburles commented:
As an alternative, if you would like to make the commits yourself. I'm happy for you to open a new PR with the changes in #89 and I'll merge that. Let me know if you have issues running the Tinytest tests, but they look good on my side.

drone1 commented:
it's okay, you can make the commits, but you might consider removing the
dependency on underscore as well.

On Wed, Nov 27, 2024 at 9:36 PM David Burles ***@***.***>
wrote:

> As an alternative, if you would like to make the commits yourself. I'm
> happy for you to open a new PR with the changes in #89
> <https://github.com/dburles/meteor-collection-helpers/pull/89> and I'll
> merge that. Let me know if you have issues running the Tinytest tests, but
> they look good on my side.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#issuecomment-2504730704>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPF4DOBH5N7UDV3T7GD2CYULXAVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBUG4ZTANZQGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dburles commented:
Thanks @drone1. Yeah that's definitely a plan. I'm just not sure if it will break Meteor 1 and 2 support support if we switch to `Object.keys`/`forEach`.

dburles commented:
Published as 2.0.0! Thanks for all your effort @drone1 and @harryadel for the test workflow!

drone1 commented:
OK great. Thanks

On Mon, Dec 9, 2024 at 10:03 PM David Burles ***@***.***>
wrote:

> Published as 2.0.0! Thanks for all your effort @drone1
> <https://github.com/drone1> and @harryadel <https://github.com/harryadel>
> for the test workflow!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dburles/meteor-collection-helpers/pull/88#issuecomment-2529441946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQIEPCVUS6TT2QOP424SUL2EYAS3AVCNFSM6AAAAABPA7LCPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMRZGQ2DCOJUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
